Feature/performance

dephillipsmichael opened the pull request:
Added performance screen to BridgeAppSDK in a way that is easy to override the UI and functionality of the ViewController

syoung-smallwisdom commented on BridgeAppSDK/SBAAppDelegate.swift:
```diff
@@ -83,6 +83,11 @@ public let SBAMainStoryboardName = "Main"
     }
     
+    open var studyDuration: TimeInterval {
+        let oneYear = TimeInterval(60 * 60 * 24 * 365)
```
Use a calendar. This works for 2017 but not 2018. ;)  Also, I'd suggest changing this to DateComponents instead since the rule might be "3 months" etc.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceDataSource.swift:
```diff
@@ -0,0 +1,169 @@
+    case completedActivities
+}
+
+open class SBAPerformanceDataSource {
```
I don't really see the value that this class is adding. It's not really a data source b/c it's not a protocol with a concrete implementation.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+    
+    open func registerCells() {
+        // Register UITableViewCells
+        for cell in blankPerformanceCells {
```
blankPerformanceCells?

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        default: break
+        }
+        
+//        cell.layoutIfNeeded() // <- added to fix bug where user had to scroll before correct autolayout height was calculated
```
Then why is it commented out? ;)

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        return _activityManager
+    }
+    lazy fileprivate var _activityManager : SBABaseScheduledActivityManager = {
+        return SBAScheduledActivityManager(delegate: self)
```
Why not use the base class for the default implementation since you don't gain anything from this.  Also, if you refactor `SBABaseScheduledActivityManager` to use a a start/end date rather than daysBehind / daysAhead (deprecate those?) then you can set it up to show you the whole past or a range, as needed.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+
+import Foundation
+
+open class SBAPerformanceViewController: UITableViewController, SBAScheduledActivityManagerDelegate {
```
I think a better implementation of the architecture for this would be to use a data source that is a protocol with a general-use implementation (extension on  `SBABaseScheduledActivityManager`?). Otherwise, for expediency in getting this to production, you could just get ride of the data source and implement as a part of this controller.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        }
+        
+        // Register UITableViewHeaderFooterViews
+        self.tableView.register(UINib.init(nibName: SBAPerformanceSectionHeader.staticCellName, bundle: Bundle(for: SBAPerformanceSectionHeader.self)), forHeaderFooterViewReuseIdentifier: SBAPerformanceSectionHeader.staticCellName)
```
It's one thing to include hardcoded "it only uses this set of magic values" stuff in an app. For a class that is intended to be general-purpose, this kind of thing can make overriding messy. i.e., What happens when someone wants to use a header that inherits from `SBAPerformanceSectionHeader` rather than using the header design included here?

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+                self.configureStudyProgressTimeLeftCell(cell: cell, for: indexPath)
+                return cell
+                
+            default: //.completedActivities
```
Use `.completedActivities` explicitly.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        }
+        
+        switch sectionType {
+        case .studyProgress:
```
Break this into a separate method. First, large blocks of nested code is difficult to read (and easy to mess up later). Second, that will allow override.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+            max = 200
+        } else if (average > 200) {
+            max = average
+        }
```
Um, no. Create a struct with min/max/average and return it in a method that calculates those values for a given activity.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        }
+        
+        cell.leftLegendLabel.text = "\(min)"
+        cell.rightLegendLabel.text = "\(max)"
```
I take that back. Also, the struct should return a formatter that is appropriate to the data.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+     */
+    open func configureStudyProgressTimeLeftCell(cell: SBAPerformanceProgressLinePlotTableViewCell, for indexPath: IndexPath)
+    {
+        let appDelegate = UIApplication.shared.delegate as! SBAAppDelegate
```
Danger, Will Robinson! No. Don't do this, please.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        // Calculate the number of months between start and end date and set that
+        // as the line plot's progress max
+        let monthsYearsBetweenStudyDates =  calendar.dateComponents([.year, .month], from: startDate, to: endDate)
+        let lengthOfStudyInMonths = (1 + /* +1 for including current month */(monthsYearsBetweenStudyDates.month ?? 0)) + ((monthsYearsBetweenStudyDates.year ?? 0) * 12) // 12 months in a year
```
Ok, first, if you are using the current calendar and not the Gregorian calendar then you shouldn't assume the number of months is 12. Second, just get the months. oh yeah, and if you used `DateComponents` as suggested above, this can be calculated from that.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        let midDate = calendar.date(byAdding: .month, value: lengthOfStudyInMonths / 2, to: startDate)!
+        cell.midLegendLabel.text = monthStrs[(calendar.dateComponents([.month], from: midDate).month! - 1)] // -1 for 0 array indexing
+        let month = calendar.dateComponents([.month], from: endDate).month!
+        cell.rightLegendLabel.text = monthStrs[month - 1] // - 1 for 0 array indexing
```
````
let monthFormatter = DateFormatter()
monthFormatter.setLocalizedDateFormatFromTemplate("MMM")
cell.leftLegendLabel.text = monthFormatter.string(from: startDate)
let midDate = calendar.date(byAdding: .month, value: lengthOfStudyInMonths / 2, to: startDate)!
cell.midLegendLabel.text = monthFormatter.string(from: midDate)
cell.rightLegendLabel.text = monthFormatter.string(from: endDate)
````

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        cell.linePlotView.progressMax = lengthOfStudyInMonths
+        
+        // Apply the value of 0 to 1 to the study's duration
+        cell.linePlotView.progress = Int(Float(lengthOfStudyInMonths) * normalizedValue)
```
Your comment and the "progress" don't match. There are no `Int` values between 0 and 1.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+    open func attributedTimeLeftStr() -> NSMutableAttributedString {
+        let formatter = DateComponentsFormatter()
+        formatter.allowedUnits = [.month, .day]
+        formatter.unitsStyle = .full
```
Where is this used?

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        let (months, days) = timeUntilStudyEnds()
+        dateComponents.month = months
+        dateComponents.day = days
+        let timeUntilStudyEndsStr = String(format: "%d MONTHS %d DAYS", months, days)
```
localize

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        var dateComponents = DateComponents()
+        let (months, days) = timeUntilStudyEnds()
+        dateComponents.month = months
+        dateComponents.day = days
```
Not seeing where the date components is being used, either.

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+        attributedString.setAttributes(detailNumberAttributes(), range: monthsRange)
+        let daysRange = NSString(string: timeUntilStudyEndsStr).range(of: String(days))
+        attributedString.setAttributes(detailNumberAttributes(), range: daysRange)
+        return attributedString
```
Localize.

````
let formatter = DateComponentsFormatter()
formatter.maximumUnitCount = 2
formatter.unitsStyle = .full
formatter.allowedUnits = [.year, .month, .day]
guard let timeUntilStudyEndsStr = formatter.string(from: Date(), to: endDate)?.localizedUppercase 
else { 
    return 
}

let attributedString = NSMutableAttributedString(
    string: timeUntilStudyEndsStr,
    attributes: detailTextAttributes())

var startIndex = 0
let searchString = NSString(string: timeUntilStudyEndsStr)
let length = searchString.length
while startIndex != NSNotFound, startIndex < length {
    let range = searchString.rangeOfCharacter(from: .decimalDigits, options: [], range: NSMakeRange(startIndex, length - startIndex))
    startIndex = range.location + range.length
    if startIndex != NSNotFound {
        attributedString.setAttributes(detailNumberAttributes(), range: range)
    }
}
````

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+                // Right now this is a bit of a workaround to generate a score for activities that don't
+                // have them yet
+                // easieg 6/3/17
+                scores.append(0)
```
Please change the comment

syoung-smallwisdom commented on BridgeAppSDK/SBAPerformanceViewController.swift:
```diff
@@ -0,0 +1,562 @@
+            endDateWithCurrentMonth = calendar.date(byAdding: dateComponents as DateComponents, to: endDate)
+            dayDif = calendar.dateComponents([.day], from: now, to: endDateWithCurrentMonth!)
+        }
+        return (monthDif.month!, dayDif.day!)
```
tinker with Playground. This is seriously obfuscated.

dephillipsmichael commented:
I agree, at this point I should just add these classes into Journey Pro and continue to iterate on them to make them proper for BASDK.   But I'd like to keep this PR open for addressing the comments.

syoung-smallwisdom commented:
groovy
